src/main/java/org/zalando/nakadi/domain/BatchItem.java

```diff
@@ -64,15 +68,29 @@ public static EmptyInjectionConfiguration build(final int position, final boolea
     private int eventSize;
 
     public BatchItem(
-            final String event,
+            final String rawEvent,
             final boolean strictParsing,
             final EmptyInjectionConfiguration emptyInjectionConfiguration,
             final InjectionConfiguration[] injections,
             final List<Integer> skipCharacters) {
-        this.rawEvent = event;
+        this.rawEvent = rawEvent;
         this.skipCharacters = skipCharacters;
-        this.event = strictParsing ? StrictJsonParser.parseObject(event) : new JSONObject(event);
-        this.eventSize = event.getBytes(StandardCharsets.UTF_8).length;
+        this.event = new JSONObject(rawEvent);
+
+        // We will first release in a shadow mode when the strict parser does second parsing
+        // and we compare if the result is the same as with the old parser; we also log invalid events;
+        if (strictParsing) {
+            try {
+                final JSONObject shadowEvent = StrictJsonParser.parseObject(rawEvent);
+                if (!shadowEvent.toString().equals(this.event.toString())) {
```

Comment:
@v-stepanov This `toString` operation is very expensive and is the core of the optimization that @antban did that modified the behaviour couple of months ago. I think we cannot go this far. The best we can do is to verify if it throws exception or not.

Reply:
Then we will not be able to validate the correctness. We can check the performance on staging and cancel if it consumes too much resources.

Reply:
ok, I get it. Then we need to be mindful and run it for a limited amount of time. 1 day I would say.